### DISCUSSION TITLE
Preserve extra targets when running ns clean

### DISCUSSION THREAD

**Discussion by ericuldall** [NativeScript/NativeScript#10370] [Q&A]
I have created a new watchOS target associated with my app from xcode. Sometimes I need to run `ns clean` but doing so deletes my watchOS target.

How do I configure the nativscript build to preserve my watchOS code?

**Comment by rigor789** [comment]
@ericuldall you can place your extra targets in `App_Resources/iOS/extensions` and define the build properties in the `extension.json` - this will make it so after `ns clean` the target gets re-added automatically. The details are currently not well documented - we've recently had to deal with a similar setup, and will try to document extensions/targets very soon...

```
App_Resources/iOS/
  extensions/
    TestWatchOS/
      ...
      Info.plist
      extension.json
```

```jsonc
// extension.json
{
  "frameworks": [
    "WatchKit"
  ],
  "targetBuildConfigurationProperties": {
    "DEVELOPMENT_TEAM": "\"XXXXXX\"",
    "CODE_SIGN_ENTITLEMENTS": "\"$(SRCROOT)/../../App_Resources/iOS/extensions/TestWatchOS/TestWatchOS.entitlements\"",
    "CODE_SIGN_STYLE": "Automatic",
​
    "PRODUCT_BUNDLE_IDENTIFIER": "org.nativescript.app.TestWatchOS",
​
    "CLANG_ANALYZER_NONNULL": "YES",
    "CLANG_ANALYZER_NUMBER_OBJECT_CONVERSION": "YES_AGGRESSIVE",
    "CLANG_CXX_LANGUAGE_STANDARD": "\"gnu++20\"",
    "CLANG_ENABLE_OBJC_WEAK": "YES",
    "CLANG_WARN_DOCUMENTATION_COMMENTS": "YES",
    "CLANG_WARN_UNGUARDED_AVAILABILITY": "YES_AGGRESSIVE",
    "COPY_PHASE_STRIP": "NO",
    "CURRENT_PROJECT_VERSION": "1",
    "DEBUG_INFORMATION_FORMAT": "\"dwarf-with-dsym\"",
    "GCC_C_LANGUAGE_STANDARD": "gnu11",
    "GCC_WARN_UNINITIALIZED_AUTOS": "YES_AGGRESSIVE",
    "GENERATE_INFOPLIST_FILE": "YES",
    "INFOPLIST_FILE": "../../App_Resources/iOS/extensions/TestWatchOS/Info.plist",
    "INFOPLIST_KEY_CFBundleDisplayName": "TestWatchOS",
    "INFOPLIST_KEY_NSHumanReadableCopyright": "\"Copyright  2023 NativeScript. All rights reserved.\"",
    "IPHONEOS_DEPLOYMENT_TARGET": "16.4",
    "MARKETING_VERSION": "1.0",
    "MTL_ENABLE_DEBUG_INFO": "NO",
    "MTL_FAST_MATH": "YES",
    "PRODUCT_NAME": "\"$(TARGET_NAME)\"",
    "SKIP_INSTALL": "YES",
    "SUPPORTS_MACCATALYST": "YES",
    "SWIFT_COMPILATION_MODE": "wholemodule",
    "SWIFT_EMIT_LOC_STRINGS": "YES",
    "SWIFT_OPTIMIZATION_LEVEL": "\"-O\"",
    "SWIFT_VERSION": "5.0",
    "TARGETED_DEVICE_FAMILY": "\"1,2\""
  }
}
```

> **Note**: this is an example `extension.json` - the best way to get the correct values is to:
1. open the project in XCode
2. add your extension
3. find the new target inside `platforms/ios/.../yourproject.pbproject`
4. manually convert those build flags to the above JSON format

---

We also plan on improving the dx/flow of this in the future, though initially we'll just focus on documenting this in detail, given it's a one-time setup thing in most cases.

**Reply by ericuldall** [reply]
Thanks for the updates. I'll try this. I'm pretty green when it comes to iOS dev so i'm not familiar with all the important details in those files. I'll do my best to piece it together from your notes above.

**Reply by rigor789** [reply]
Hey @ericuldall, just found this in the old docs: https://v7.docs.nativescript.org/tooling/ios-watch-apps

This should still work - mostly same steps as outlined above, but the directory would be `App_Resources/iOS/watchapp` and the json file would be `watchapp.json`
